### PR TITLE
User Story update + payment state engine

### PR DESCRIPTION
**Pull request by bigludo7**
Update of the user story (US) to have separate one for  1-step & 2 steps + US for retrieve by it and retrieve by list. Adding a proposal for Payment state engine

### CODE REVIEW

**Review by alabajnaid — DISMISSED** [review]

File: documentation/API_documentation/Carrier Billing User Story.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+**User Story: Make a Payment request in one step**
+<br>
+
+| **Item** | **Details** |
+| ---- | ------- |
+| ***Summary*** | As an application developer belonging to an enterprise, I want to request (using my application server/backend service) a payment (one-time fee -in initial phase- or recurring -in a later phase- ) to be billed against a customer's carrier bill, so that I can offer customers payment alternatives.  |
+| ***Roles, Actors and Scope*** | **Roles:** Customer:User<br> **Actors:** Application service providers, hyperscalers, application developers. <br> **Scope:** *Request Payment/Check out on Carrier Bill* \- Initiate Carrier Billing approval process with Customer from Carrier Channels (SMS/App) |
+| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has Carrier Billing activated as payment method.</li><li>The Customer:User have initiated payment via Carrier Billing/Checkout from application/portal/product channel</li><li>The PaymentRequester have verified that associated carrier service belongs to the Customer:User (Generally as part of adding a new payment method, 2-Factor-Authorization or equivelant)</li></ol>|
```
Comment (PedroDiez) [inline documentation/API_documentation/Carrier Billing User Story.md:8]:
```suggestion
| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has Carrier Billing activated as payment method.</li><li>The Customer:User has initiated payment via Carrier Billing/Checkout from application/portal/product channel</li><li>The PaymentRequester has verified that associated carrier service belongs to the Customer:User (Generally as part of adding a new payment method, 2-Factor-Authorization or equivalent)</li></ol>|

```

File: documentation/API_documentation/Carrier Billing User Story.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+| ***Roles, Actors and Scope*** | **Roles:** Customer:User<br> **Actors:** Application service providers, hyperscalers, application developers. <br> **Scope:** *Request Payment/Check out on Carrier Bill* \- Initiate Carrier Billing 2-steps approval process with Customer from Carrier Channels (SMS/App) |
+| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has Carrier Billing activated as payment method.</li><li>The Customer:User have initiated 2-steps payment via Carrier Billing/Checkout from application/portal/product channel</li><li>The PaymentRequester have verified that associated carrier service belongs to the Customer:User (Generally as part of adding a new payment method, 2FA or equivelant)</li></ol>|
```
Comment (PedroDiez) [inline documentation/API_documentation/Carrier Billing User Story.md:20]:
```suggestion
| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has Carrier Billing activated as payment method.</li><li>The Customer:User has initiated 2-steps payment via Carrier Billing/Checkout from application/portal/product channel</li><li>The PaymentRequester has verified that associated carrier service belongs to the Customer:User (Generally as part of adding a new payment method, 2FA or equivalent)</li></ol>|

```

File: documentation/API_documentation/Carrier Billing User Story.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+| ***Activities/Steps*** | **Starts when:** The customer application server makes a request to the Carrier Billing Checkout to trigger a 2-steps payment. <br>**Intermediate steps:** The customer application server makes a request to confirm or cancel the payment preparation <br>**Ends when:** If payment confirmation is provided the payment is completed. If a payment cancellation is provided (explicit cancellation) or if neither a confirmation or cancellation is received after a predefined business delay (implicitly by expiration), the payment preparation is cancelled. <br>For both case the customer application customer can retrieve and/or query details of that payment<br> |
```
Comment (PedroDiez) [inline documentation/API_documentation/Carrier Billing User Story.md:21]:
```suggestion
| ***Activities/Steps*** | **Starts when:** The customer application server makes a request to the Carrier Billing Checkout to trigger a 2-steps payment. <br>**Intermediate steps:** The customer application server makes a request to confirm or cancel the payment preparation <br>**Ends when:** If payment confirmation is provided the payment is completed. If a payment cancellation is provided (explicit cancellation) or if neither a confirmation or cancellation is received after a predefined business delay (implicitly by expiration), the payment preparation is cancelled. <br>For both cases the customer application can retrieve and/or query details of that payment<br> |

```

File: documentation/API_documentation/Carrier Billing User Story.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+| ***Post-conditions*** | After customer application server performs the payment, payment details can be queried. A payment can stay in preparation status only for a predefined business delay.  |
+| ***Exceptions*** | Several exceptions might occur during the Carrier Billing API operations:<br><ul><li>Unauthorized: Not valid credentials (e.g. use of already expired access token).</li><li>Invalid input: Not valid input data to invoke operation like for example: <br> payment identifier not existing or not linked to the customer:user,<br>payment confirmation fo an expired payment,<br>payment cancellation request for an already confirmed payment).</li><li>Denied by Carrier: Any user/business condition and/or regulation that forbids the perform of the payment. Payment preparation as payment confirmation could be denied</li><li>Conflict: Internal configuration policies don't allow for Payment Execution/Cancellation.</li></ul>|
```
Comment (PedroDiez) [inline documentation/API_documentation/Carrier Billing User Story.md:23]:
```suggestion
| ***Exceptions*** | Several exceptions might occur during the Carrier Billing API operations:<br><ul><li>Unauthorized: Not valid credentials (e.g. use of already expired access token).</li><li>Invalid input: Not valid input data to invoke operation like for example: <br>merchant information not provided or unknown,<br>payment identifier not existing or not linked to the customer:user,<br>payment confirmation for an expired payment,<br>payment cancellation request for an already confirmed payment).</li><li>Denied by Carrier: Any user/business condition and/or regulation that forbids the perform of the payment. Payment preparation as payment confirmation could be denied</li><li>Conflict: Internal configuration policies don't allow for Payment Execution/Cancellation.</li></ul>|

```

File: documentation/API_documentation/Carrier Billing User Story.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+<br>
+
+**User Story: Retrieve a Payment list**
+<br>
+
+| **Item** | **Details** |
+| ---- | ------- |
+| ***Summary*** | As an application developer belonging to an enterprise, I want to get a list of performed payments from criteria (using my application server/backend service), by means of that application.  |
+| ***Roles, Actors and Scope*** | **Roles:** Customer:User<br> **Actors:** Application service providers, hyperscalers, application developers. <br> **Scope:** *Request Payments list done against Carrier Billing* |
+| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has performed payments via application, both success and failed</li><li>The requester provide a list of criteria(s) to select payment (Customer/user identifier, payment status)</li></ol>|
```
Comment (PedroDiez) [inline documentation/API_documentation/Carrier Billing User Story.md:46]:
```suggestion
| ***Pre-conditions*** |The preconditions are listed below:<br><ol><li>The Customer:User has performed payments via application, both success and failed</li><li>The requester provides a list of criteria(s) to select payment (Customer/user identifier, payment status)</li></ol>|

```

File: documentation/API_documentation/Payment lifecycle source.puml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+state choice <<choice>>
+state end3    <<end>>
+
+[*] --> choice
+choice --> Processing : async
+
+
+choice --> fork : sync
+
+note left of choice: Business response is provided\nsynchronously or synchronously
```
Comment (PedroDiez) [inline documentation/API_documentation/Payment lifecycle source.puml:13]:
```suggestion
note left of choice: Business response is provided\nsynchronously or asynchronously

```

**Review by PedroDiez — COMMENTED** [review]
LGTM in advance

Some minor editorials

**Review by PedroDiez — APPROVED** [review]
LGTM

**Review by PedroDiez — APPROVED** [review]
LGTM

I directly committed to have PR updated and merge it. As we gave already ok
@bigludo7, @alabajnaid

**Pull request merged**